In-source zinc's LM integration code

dwijnand opened the pull request:
This requires a new version of Zinc in order to build (tested using local source dependencies).

harrah committed:
reorganize compilation modules

harrah committed:
implement embedded interface

gkossakowski committed:
Follow source layout convention supported by Eclipse.

Moved source files so directory structure follow package
structure. That makes it possible to use Scala Eclipse plugin
with sbt's source code.

eed3si9n committed:
added scalariform

Duhemm committed:
Find most specific version of compiler interface sources

This commit introduces a mechanism that allows sbt to find the most
specific version of the compiler interface sources that exists using
Ivy.

For instance, when asked for a compiler interface for Scala 2.11.8-M2,
sbt will look for sources for:

 - 2.11.8-M2 ;
 - 2.11.8 ;
 - 2.11 ;
 - the default sources.

This commit also modifies the build definition by removing the
precompiled projects and configuring the compiler-interface project so
that it publishes its source artifacts in a Maven-friendly format.

pdalpra committed:
Replace procedure syntax by explicit Unit annotation

eed3si9n committed:
moved subprojects around

Duhemm committed:
Hide Ivy logs in compiler-interface sources retrieval

Because in most cases there aren't version-specific sources, we expect
the retrieval to fail a number of times before succeeding. This
generates a lot of noise in sbt's log, so the logs will now be shown if
and only all the versions fail.

Duhemm committed:
Download compiler interface sources in boot directory

Duhemm committed:
Update to sbt/io 1.0.0-M1, fix dependencies

Duhemm committed:
Add recommended compiler flags, suppress a lot of warnings

Duhemm committed:
Fix error in migration to sbt/io

Duhemm committed:
Update sbt modules, migrate to scalatest

 - Update sbt/util to 1.0.0-M3
 - Update sbt/librarymanagement to 1.0.0-M2

Also, migrate the tests from specs2 to scalatest.

eed3si9n committed:
Merge pull request #3 from Duhemm/compiler-interface-logging

Compiler interface logging (sbt/sbt#2177)

eed3si9n committed:
Fixes build

Duhemm committed:
Retrieve version of compiler interface from properties file

Previously we used the version of the component manager. However, these
two versions do not need to be in sync, which would make us try to
retrieve the wrong version of the compiler interface sources.

Duhemm committed:
Specify the compiler bridge module in a setting

In order to restore reproducibility of builds, we no longer cascade over
the possibly available versions of the compiler bridge sources (a
specific version of the bridge sources may not be available one day, but
exist on the next day), but rather let the build definition configure
which module to use.

Fixes sbt/sbt#2196

Duhemm committed:
Retrieve compiler bridge sources in temp directory

Note that they won't be downloaded again, because the component compiler
will look for a previously-compiled version of the compiler bridge
before trying to fetch the sources again. If they've already been
downlaoded, then they have been compiled and a compiled version of the
compiler bridge already exists.

Duhemm committed:
Rename binary form of compiler bridge

This will avoid all clashes between modules that may have the same name
as other components of sbt, or two different compiler bridges that would
happen to have the same name.

Duhemm committed:
Fix deprecation message

Duhemm committed:
Retrieve java compiler interfaces using Ivy

Previously the component compiler depended on a `componentID` added to
some interfaces to be able to compile the compiler interface. sbt's
launcher placed artifacts marked with this `componentID` in a special
directory that was then used by the ComponentCompiler to be able to
compile the compiler interface.

This commit modifies the ComponentCompiler so that it does not expect
these components to be already present in some directory, but instead
downloads it using Ivy when required.

eed3si9n committed:
Reorganize subprojects around

eed3si9n committed:
New scalariform

eed3si9n committed:
Fix the hardcoded ModuleID to compiler-interface

eed3si9n committed:
Move things around further.

Duhemm committed:
Put `util-interface` on classpath when compiling compiler bridge

Also, write tests to make sure that we remain able to compile the
compiler bridge for different versions of Scala.

Fixes sbt/incrementalcompiler#32

Duhemm committed:
Fix tests that fetch and build the compiler bridge.

Duhemm committed:
Fix command and definition of compiler bridge

The compiler bridge had a dependency on the scala compiler in the
version defined in the build definition. This means that when trying to
fetch the compiler bridge for, say, Scala 2.9.2, we would also pull in
the Scala compiler 2.10.5.

The compiler bridge now has a "provided" dependency on the Scala
compiler.

eed3si9n committed:
Quieter log during scripted

eed3si9n committed:
Use real local to retrieve compiler bridge during test. Fixes #36

eed3si9n committed:
librarymanagement 0.1.0-M7

Duhemm committed:
Fix compiler bridge sources retrieval

gkossakowski committed:
Zinc is compatible with Scala 2.10 and 2.11 only

The class-based-dependencies branch nuked compatibility with older Scala
compiler versions.

gkossakowski committed:
Rename incrementalcompiler to zinc

Reflect the rename of the repo to zinc in project's structure.

eed3si9n committed:
Update dependencies

eed3si9n committed:
Don't use other projects' tests

Duhemm committed:
Generate `ClasspathOptions` with sbt-datatype

Also rename the previous object called `ClasspathOptions` to
`ClasspathOptionsUtil`.

eed3si9n committed:
Merge pull request #128 from Duhemm/wip/compiler-api

Minimal API for `ScalaCompiler`, expose in `Compilers`

kiritsuku committed:
Rename CompilerInterfaceProvider to CompilerBridgeProvider

Fix #162

dwijnand committed:
Upgrade to new sbt/librarymanagement

dwijnand committed:
Bump TravisCI to 2.11.8, match crossScalaVersions

dwijnand committed:
Upgrade to newer modules, migrate to new lm api

eed3si9n committed:
Cross build to Scala 2.12

eed3si9n committed:
Switch over to using ManagedLogger

eed3si9n committed:
Migrate ComponentManager from librarymanagement

This migrates the `ComponentManager` from librarymanagement, and refactors it to clarify the double caching of the compiler bridge.

The main purpose of this refactoring is to fix sbt/sbt#2539. As it stands, becaue Ivy cache directory is being used as the secondary cache of the compiled compiler bridge, it's annoying to work with the compiler bridge. Instead of using "Ivy cache", the new `ZincComponentManager` accepts `secondaryCacheDir: Option[File]`, which during test defaults to `target/zinc-components`.

Fixes sbt/sbt#2539

eed3si9n committed:
Put copyright notice

Fixes #222

jvican committed:
Do a first pass on the Zinc internal implementation

This commit addresses several concers:

* Document better the Zinc internal implementation.
* Getting the implementation ready for moving things from internal to
  public APIs. This will be done in a follow-up commit.
* Removing `inputs` and `setups` from the `IncrementalCompilerUtil` as
  they exist mainly to interface between Scala code and Java code. The
  reality is that both Java and Scala developers can easily call the
  constructors by themselves and don't learn more API methods.
* Removes unused imports, that emit stupid warnings.
* Make code more readable, especifically in the analyzing compilers for
  both Java and Scala, which was hardly understandable and documented.
* Make `AnalyzingJavaCompiler` to extend `JavaCompiler` as
  `AnalyzingCompiler` does for `ScalaCompiler`. In an upcoming commit,
  `AnalyzingCompiler` will be renamed into `AnalyzingScalaCompiler` for
  consistency reasons.
* Replace use of `sbt.util.Logger` helpers by `sbt.util.InterfaceUtil`.

eed3si9n committed:
Adds Hello world for 2.10, 2.11, and 2.12

eed3si9n committed:
Better error message for bridge resolution

The compiler bridge dependency resolution had `missingOk = true` flag on. This would silently succeed even when one of the dependency is missing, and it would randomly fail bridge compilation, which is very confusing.

eed3si9n committed:
Bump to util 1.0.0-M22 and 1.0.0-X10

Bumping to the latest librarymanagement.

jvican committed:
Reformat zinc with scalafmt

The same as https://github.com/sbt/librarymanagement/pull/87/.

jvican committed:
Add `CompilerBridgeProvider` interface and misc.

* Add interface for the provider.
* Rename `IvyComponentCompiler` to `ZincComponentCompiler`.
* Split `ZincComponentCompiler` and `ZincComponentManager`.
* Define `IfMissing` and `InvalidComponent` into independent files.
* Rename variables and internal API to be clearer.

jvican committed:
Add headers to files that don't have them

jvican committed:
Rename `getCompilerSources` to `getCompiledBridge`

Previous name was incorrect, we actually return the compiled bridge.

jvican committed:
Fix oversight in InvalidComponent message

jvican committed:
Add `getScalaInstance` to `CompilerBridgeProvider`

jvican committed:
Remove custom `getScalaInstance` and use default

jvican committed:
Test `getScalaInstance` & `getCompiledBridge`

* Remove unused code in `BridgeProviderSpecification` and test the real
  code in the default provider.
* Don't use temporary directory to download the Scala jars, retrive them
  in a directory specified by the user.
* Use `java.net.ClassLoader` instead of `ScalaClassLoader`.
* Use the `ScalaInstance` interface, not the implementation. Remove any
  reference to the implementation.

jvican committed:
Compile bridge for 2.12.x and other Scala versions

This commit makes sure that we always check that we compile against the
latest two versions in Scala 2.10, Scala 2.11 and Scala 2.12.

This commit also removes the null loggers and uses normal loggers to
check what the output of the bridge provider is. This output is
important and the rationale to do this is that it should be visible for
us in order to detect cases where there's a regression or the output is
modified in some way. Output is short anyway.

jvican committed:
Fix error when showing unresolved warning lines

jvican committed:
Add Java API for Zinc Ivy integration

* Adds a method to fetch the default implementation of the compiler
  provider.
* Adds a method to get a default `IvyConfiguration` from a small number
  of parameters.

The public API is implemented in Java to ensure binary compatibility no
matter what happens to the internal Scala implementations. It also makes
it easier for Java users to use the API and avoid importing MODULE$'s
and Scala generated class files.'

jvican committed:
Add public method to get a local ivy resolver

jvican committed:
Add public method to get no-op global lock

jvican committed:
Remove verbose Java global lock implementation

jvican committed:
Add public method to get default component provider

Expose default component provider through a static method in
`ZincCompilerBridge`. This method is necessary to invoke the bridge.

Removes the previous global lock and component provider implementations
in `BridgeProviderSpecification`.

jvican committed:
Remove the use of `BaseIvySpecification`

This is the proof that we're now testing the Zinc public API because
we're resolving with the ivy options and resolution specified in the
default compiler bridge provider.

jvican committed:
Add `ZincCompilerUtil` to get the Scalac compiler

* Move `IncrementalCompilerUtil` to `ZincCompilerUtil`.
* Change `IncrementalCompilerUtil` to an interface with static methods.
* Remove bug in which ScalaInstance implementation was leaking to the
  public API.

jvican committed:
Hide Zinc Compilation implementation

Before, we had made this public in `sbt.inc`, but this is not necessary
anymore.

jvican committed:
Move ZincBridgeProvider to `xsbti.compile`

Before, we were using `sbt.inc`. As we have settled to define all the
interfaces in the `xsbti.compile` namespaces, it's not worth it to force
users remember other locations for the interfaces forwarding the default
implementations in Scala.

jvican committed:
Move `ClasspathOptionsUtil` to the Java public API

Move the previous Scala implementation to Java as static methods.

jvican committed:
Add headers to files missing it

jvican committed:
Rename `getXXX` by `fetchXXX`

As suggested by Eugene.

eed3si9n committed:
Bump all the modules

eed3si9n committed:
Add scalaCompiler variant that accepts ModuleID

eed3si9n committed:
Fix ScalaFmt wiring

Duhemm committed:
Adapt Zinc to use the new LibraryManagement API

jvican committed:
Remove any reference to `F0` and `F1`

Syncs up with https://github.com/sbt/util/pull/84.

eed3si9n committed:
Merge branch 'pr/335' into wip/integration

# Conflicts:
#	internal/zinc-ivy-integration/src/main/scala/sbt/internal/inc/ZincComponentCompiler.scala
#	project/Dependencies.scala

jvican committed:
Use Scala 2.12.3

We have to look into the reason why the shaded version of the compiler
plugin reports the following error and halts the build:

```
protoc-jar: executing: [/tmp/protocjar2818613521400194680/bin/protoc.exe, --plugin=protoc-gen-scala=/tmp/protocbridge4921434211196915044, --scala_out=grpc:/data/rw/code/scala/zinc/internal/zinc-persist/target/scala-2.12/src_managed/main, -I/data/rw/code/scala/zinc/internal/zinc-persist/src/main/protobuf, -I/data/rw/code/scala/zinc/internal/zinc-persist/target/protobuf_external, /data/rw/code/scala/zinc/internal/zinc-persist/src/main/protobuf/schema.proto]
java.lang.AbstractMethodError: Method scalapb/ScalaPbCodeGenerator$.run(Lcom/google/protobuf/compiler/PluginProtos$CodeGeneratorRequest;)Lcom/google/protobuf/compiler/PluginProtos$CodeGeneratorResponse; is abstract
	at scalapb.ScalaPbCodeGenerator$.run(ScalaPbCodeGenerator.scala)
	at protocbridge.frontend.PluginFrontend$.$anonfun$runWithBytes$1(PluginFrontend.scala:56)
	at scala.util.Try$.apply(Try.scala:209)
	at protocbridge.frontend.PluginFrontend$.runWithBytes(PluginFrontend.scala:54)
	at protocbridge.frontend.PluginFrontend$.runWithInputStream(PluginFrontend.scala:67)
	at protocbridge.frontend.PosixPluginFrontend$.$anonfun$prepare$1(PosixPluginFrontend.scala:25)
	at scala.runtime.java8.JFunction0$mcV$sp.apply(JFunction0$mcV$sp.java:12)
	at scala.concurrent.Future$.$anonfun$apply$1(Future.scala:653)
	at scala.util.Success.$anonfun$map$1(Try.scala:251)
	at scala.util.Success.map(Try.scala:209)
	at scala.concurrent.Future.$anonfun$map$1(Future.scala:287)
	at scala.concurrent.impl.Promise.liftedTree1$1(Promise.scala:29)
	at scala.concurrent.impl.Promise.$anonfun$transform$1(Promise.scala:29)
	at scala.concurrent.impl.CallbackRunnable.run(Promise.scala:60)
	at scala.concurrent.impl.ExecutionContextImpl$AdaptedForkJoinTask.exec(ExecutionContextImpl.scala:140)
	at java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java:289)
	at java.util.concurrent.ForkJoinPool$WorkQueue.runTask(ForkJoinPool.java:1056)
	at java.util.concurrent.ForkJoinPool.runWorker(ForkJoinPool.java:1692)
	at java.util.concurrent.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:157)
```

dwijnand committed:
Add back, re-configure & re-enable Scalafmt

allantl committed:
Remove unused imports + variables

eed3si9n committed:
Split compiler bridge tests to another subproject

Splitting compiler bridge tests to another subproject because while the bridge itself can be compiled with just compiler-interface, util-interface, and Scala Compiler as dependencies, the testing introduces more (such as IO). This creates problem for new Scala versions where IO or test libraries do not exist yet (e.g. Scala 2.13.0-M2).

This also removes the Mima test due to the lack of 2.13 bridge for Zinc 1.0.0.
Compiler bridge just needs to compile itself against the interface and Scala compiler, so there's no need to run Mima test.

dwijnand committed:
Merge branch '1.0.x' into merge-1.0.x-into-1.x

* 1.0.x: (28 commits)
  Split compiler bridge tests to another subproject
  Implement compiler bridge for 2.13.0-M2
  Add yourkit acknoledgement in the README
  "sbt '++ 2.13.0-M2!' compile" does not work with sbt 1.0.0
  Add header to cached hashing spec
  Add headers to missing files
  Fix #332: Add sbt-header back to the build
  Update sbt-scalafmt to 1.12
  Make classpath hashing more lightweight
  Fix #442: Name hash of value class should include underlying type
  source-dependencies/value-class-underlying: fix test
  Ignore null in generic lambda tparams
  Improve and make scripted parallel
  Fix #436: Remove annoying log4j scripted exception
  Fix #127: Use `unexpanded` name instead of `name`
  Add pending test case for issue/127
  source-dependencies / patMat-scope workaround
  Fixes undercompilation on inheritance on same source
  Add real reproduction case for sbt/zinc#417
  Add trait-trait-212 for Scala 2.12.3
  ...

 Conflicts:
	internal/zinc-apiinfo/src/main/scala/sbt/internal/inc/ClassToAPI.scala
	project/build.properties
	zinc/src/main/scala/sbt/internal/inc/MixedAnalyzingCompiler.scala

The ClassToAPI conflict is due to:
* https://github.com/sbt/zinc/pull/393 (a 1.x PR), conflicting with
* https://github.com/sbt/zinc/pull/446 (a 1.0.x PR).

The build.properties conflict is due to different PRs bumping
sbt.version from 1.0.0 to 1.0.2 to 1.0.3. (#413, #418, #453).

The MixedAnalyzingCompiler conflict is due to:
* https://github.com/sbt/zinc/pull/427 (a 1.x PR), conflicting with
* https://github.com/sbt/zinc/pull/452 (a 1.0.x PR).

dwijnand committed:
Support scala 2.13.0-pre-* & 2.13.0-M1

Fixes sbt/sbt#3771
Fixes sbt/zinc#460

dwijnand committed:
Merge branch '1.0.x' into merge-1.0.x-into-1.1.x

* 1.0.x:
  Make PositionImpl thread-safe
  Support scala 2.13.0-pre-* & 2.13.0-M1

Commit:
Drop "final" modifier from private case classes, to prevent false positive.
Refer https://issues.scala-lang.org/browse/SI-4440

eed3si9n committed:
include only scala-library into the boot classpath during run

Fixes https://github.com/sbt/sbt/issues/3405
Ref https://github.com/scala/scala-xml/issues/195

sbt's `run` is emulated using a classloader trick that includes ScalaInstance as the parent classloader under the classpath. The problem is the ScalaInstance classloader currently contains both compiler, library, and their transitive JARs:

```scala
res0: Array[java.io.File] = Array(scala-library.jar, scala-compiler.jar, jline.jar, scala-reflect.jar, scala-xml_2.12.jar)
```

This could have been causing various issues, but most recently it showed up as wrong version of scala-xml getting prioritized over what's passed by the user.

1. new field loaderLibraryOnly is added to xsbti.ScalaInstance.
2. it is initialized to the library loader if the launcher creates it, otherwise create layered loader here.

This aims to isolate the library loader, and retain the perf.

eed3si9n committed:
Merge branch 'wip/bumpsbt' into wip/merge-1.1.x

lukaszwawrzyk committed:
Got rid of multiple warnigns

seratch committed:
Fix the several typos detected by github.com/client9/misspell

eed3si9n committed:
Refactor compiler bridge unit test

This refactors the compiler bridge unit test to use the normal Zinc facility,
namely AnalyzingCompiler that's built on Scala 2.12, but is capable of driving
the compiler bridge built on non-2.12.
This allows us to run the unit tests without any additional dependencies published for Scala 2.13.0-M5.

eed3si9n committed:
Skip Scala 2.10 and Scala 2.11 components for JDK 11

Scala 2.10 and Scala 2.11 do not support JDK 11.

eed3si9n committed:
bump Scala bridge to use 2.13.0-RC1

We are fairly certain we can maintain source-compatibility of compiler bridge from 2.13.0-RC to 2.13.x.

See also https://github.com/sbt/zinc/issues/78 :)

eed3si9n committed:
call it compiler-bridge_2.13

eed3si9n committed:
Add scala-integration resolver

dwijnand committed:
Break the dependency on LM, and thus Ivy

This patch breaks the dependency that Zinc had on sbt's Library
Management (LM) library abstraction (sbt/librarymanagement) and thus,
transitively, on Ivy.

The reason Zinc depends on LM in the first place is to dynamically
download the compiler bridge sources JAR, which it then compiles in
order to bridge from binary-compatible Zinc to the never-binary-stable
(and most of the time source-compatible) compiler API.

From my research (that is, using GitHub Search) the only user of this
feature is sbt, with all other integrations (e.g. pants, bloop, mill)
providing the compiler bridge sources JAR directly (an alternative API
entry point).

Therefore the dependency on LM and that integration code could be
untangled from Zinc and moved into sbt.  That would also give an
opportunity to move the code in LM (back) into sbt, which I also think
would be a good idea.

For now this patch leaves that integration code in the already existing
`zinc-ivy-integration` module, but which is now no longer a dependency
of any other module of Zinc, specifically it is no longer a dependency
of the `zinc` and `zincScripted` modules.  I think, though,
zinc-ivy-integration should remain in the zinc repo, as a part of the
build, being tested in CI, until it's been moved (back) into sbt.  I'd
be happy to take care of both the remaining move and removal tasks.

In order to allow removing LM, Ivy and zinc-ivy-integration as
transitive dependencies of Zinc, this patch has to make a breaking
change to the `ZincUtil` object in the zinc module.  Despite it being
defined in the `sbt.internal.inc` package (i.e. not declared as public
API) ZincUtil is in fact a used (Scala) API of Zinc (usages discovered
via GitHub Search).  Therefore I chose to leave the object in the zinc
module and only drop from the object the two methods that directly
depend on LM's API.  Specifically (1) the `getDefaultBridgeModule`
method and (2) the `scalaCompiler` overload that depends on LM were
removed.  These methods now live in a new `ZincLmUtil` object in
zinc-ivy-integration, which sbt can switch to using.

In terms of risk, from my research (GitHub Search) the only users of
those methods are sbt and bloop.  I need to confirm but I think bloop
will be able to manage this breaking change (it looks it might be just
fallback usage? TBC).  For sbt we don't actual support using new Zinc
with old sbt, so sbt can also just update its usage on its next release.
Any other usage risks being broken but given it's internal, non-public
API I think it's fair game.

The remaining changes are in the testing so that zinc-ivy-integration's
test suite continues to test its download-and-compile behaviour, while
the rest of Zinc switches to using the prebuilt compiler bridges,
reusing and generalising some pre-existing zincScripted code (as well as
re-wiring some modules) so it can be used by zinc's tests too.

dwijnand committed:
Cleanup zinc-lm-integration

dwijnand committed:
In-source zinc's LM integration code

typesafe-tools commented:
A validation involving this pull request <a href="https://jenkins.scala-sbt.org:8592/job/sbt-validator/1537/console">is in progress...</a>

typesafe-tools commented:
[//]: # (Refs: pull/4617/head-pull/658/head-develop-develop-develop)
The validator <A HREF="https://jenkins.scala-sbt.org:8592/job/sbt-validator/1537/">has checked</A> the following projects, tested using dbuild, projects built on top of each other.

Project             | Reference            | Commit
:------------------ | :------------------- | :----------------
  sbt               | <A HREF="https://github.com/sbt/sbt/pull/4617">pull/4617/head      </A> | sbt/sbt@9674e4a6e9c45eab96331a9af21715255d26179c
  zinc              | <A HREF="https://github.com/sbt/zinc/pull/658">pull/658/head       </A> | sbt/zinc@0b6470f2b8558e5c8b77e2942f02aca19e3bbad8
  io                | <A HREF="https://github.com/sbt/io/tree/develop">develop             </A> | sbt/io@ca966bcd6c73400b11c7a78e8604ea211f32e3d8
  librarymanagement | <A HREF="https://github.com/sbt/librarymanagement/tree/develop">develop             </A> | sbt/librarymanagement@f073dde812848ef443619578f55e290f3863bcaa
  util              | <A HREF="https://github.com/sbt/util/tree/develop">develop             </A> | sbt/util@473d42ac782f9449dbd7c7940a67208abbff1861

:x: The result is: ***FAILED***
<A HREF="https://jenkins.scala-sbt.org:8592/job/sbt-validator/parambuild?sbt=pull/4617/head&zinc=pull/658/head&io=develop&librarymanagement=develop&util=develop&deploy_to_bintray=false&run_sbt_community=false">(restart)</A>

typesafe-tools commented:
A validation involving this pull request <a href="https://jenkins.scala-sbt.org:8592/job/sbt-validator/1538/console">is in progress...</a>

eed3si9n commented:
Could we keep git history somehow?

dwijnand commented:
Sure, I'll do the same thing I did in https://github.com/sbt/sbt/pull/3302.

I'll probably get round to it in a few days.

typesafe-tools commented:
[//]: # (Refs: pull/4617/head-pull/658/head-develop-develop-develop)
The validator <A HREF="https://jenkins.scala-sbt.org:8592/job/sbt-validator/1538/">has checked</A> the following projects, tested using dbuild, projects built on top of each other.

Project             | Reference            | Commit
:------------------ | :------------------- | :----------------
  sbt               | <A HREF="https://github.com/sbt/sbt/pull/4617">pull/4617/head      </A> | sbt/sbt@7466ee4913d5281b99130c217c7bac02347adb76
  zinc              | <A HREF="https://github.com/sbt/zinc/pull/658">pull/658/head       </A> | sbt/zinc@0b6470f2b8558e5c8b77e2942f02aca19e3bbad8
  io                | <A HREF="https://github.com/sbt/io/tree/develop">develop             </A> | sbt/io@ca966bcd6c73400b11c7a78e8604ea211f32e3d8
  librarymanagement | <A HREF="https://github.com/sbt/librarymanagement/tree/develop">develop             </A> | sbt/librarymanagement@f073dde812848ef443619578f55e290f3863bcaa
  util              | <A HREF="https://github.com/sbt/util/tree/develop">develop             </A> | sbt/util@473d42ac782f9449dbd7c7940a67208abbff1861

:x: The result is: ***FAILED***
<A HREF="https://jenkins.scala-sbt.org:8592/job/sbt-validator/parambuild?sbt=pull/4617/head&zinc=pull/658/head&io=develop&librarymanagement=develop&util=develop&deploy_to_bintray=false&run_sbt_community=false">(restart)</A>

typesafe-tools commented:
A validation involving this pull request <a href="https://jenkins.scala-sbt.org:8592/job/sbt-validator/1539/console">is in progress...</a>

typesafe-tools commented:
[//]: # (Refs: pull/4617/head-pull/658/head-develop-develop-develop)
The validator <A HREF="https://jenkins.scala-sbt.org:8592/job/sbt-validator/1539/">has checked</A> the following projects, tested using dbuild, projects built on top of each other.

Project             | Reference            | Commit
:------------------ | :------------------- | :----------------
  sbt               | <A HREF="https://github.com/sbt/sbt/pull/4617">pull/4617/head      </A> | sbt/sbt@7466ee4913d5281b99130c217c7bac02347adb76
  zinc              | <A HREF="https://github.com/sbt/zinc/pull/658">pull/658/head       </A> | sbt/zinc@0b6470f2b8558e5c8b77e2942f02aca19e3bbad8
  io                | <A HREF="https://github.com/sbt/io/tree/develop">develop             </A> | sbt/io@ca966bcd6c73400b11c7a78e8604ea211f32e3d8
  librarymanagement | <A HREF="https://github.com/sbt/librarymanagement/tree/develop">develop             </A> | sbt/librarymanagement@f073dde812848ef443619578f55e290f3863bcaa
  util              | <A HREF="https://github.com/sbt/util/tree/develop">develop             </A> | sbt/util@473d42ac782f9449dbd7c7940a67208abbff1861

:x: The result is: ***FAILED***
<A HREF="https://jenkins.scala-sbt.org:8592/job/sbt-validator/parambuild?sbt=pull/4617/head&zinc=pull/658/head&io=develop&librarymanagement=develop&util=develop&deploy_to_bintray=false&run_sbt_community=false">(restart)</A>

dwijnand commented:
```
[sbt212] [warn] 	::::::::::::::::::::::::::::::::::::::::::::::
[sbt212] [warn] 	::          UNRESOLVED DEPENDENCIES         ::
[sbt212] [warn] 	::::::::::::::::::::::::::::::::::::::::::::::
[sbt212] [warn] 	:: org.scala-sbt#compiler-bridge_2.13.0-M5;1.3.0-bin-20190418T141950: not found
[sbt212] [warn] 	::::::::::::::::::::::::::::::::::::::::::::::
[sbt212] [info] ZincComponentCompilerSpec:
[sbt212] [info] - should compile the bridge for Scala 2.10.5 and 2.10.6 *** FAILED ***
[sbt212] [info]   sbt.internal.inc.InvalidComponent: The compiler bridge sources org.scala-sbt:compiler-bridge_2.10:1.3.0-bin-20190418T141950:compile could not be retrieved.
```

*head scratch*

eed3si9n commented:
Is the bridge published?

dwijnand commented:
I guess not?  The head scratching comes from why it wants the 2.13 bridge in the 2.10 test case.  I'll iterate locally and fix the tests (I only iterated on tests compiling).

typesafe-tools commented:
A validation involving this pull request <a href="https://jenkins.scala-sbt.org:8592/job/sbt-validator/1554/console">is in progress...</a>

typesafe-tools commented:
[//]: # (Refs: pull/4617/head-develop-develop-develop-develop)
The validator <A HREF="https://jenkins.scala-sbt.org:8592/job/sbt-validator/1554/">has checked</A> the following projects, tested using dbuild, projects built on top of each other.

Project             | Reference            | Commit
:------------------ | :------------------- | :----------------
  sbt               | <A HREF="https://github.com/sbt/sbt/pull/4617">pull/4617/head      </A> | sbt/sbt@044ced544fa2422fa4761efccff17f9b207db756
  zinc              | <A HREF="https://github.com/sbt/zinc/tree/develop">develop             </A> | sbt/zinc@a5c269c729100140981e1c9ea49568a256fa92f2
  io                | <A HREF="https://github.com/sbt/io/tree/develop">develop             </A> | sbt/io@dda58a019335c1d8480a516bdd1e97ede0b7bdcb
  librarymanagement | <A HREF="https://github.com/sbt/librarymanagement/tree/develop">develop             </A> | sbt/librarymanagement@f073dde812848ef443619578f55e290f3863bcaa
  util              | <A HREF="https://github.com/sbt/util/tree/develop">develop             </A> | sbt/util@3568b13a2e257f33957dd32c51c56ce9c378c9f9

:x: The result is: ***FAILED***
<A HREF="https://jenkins.scala-sbt.org:8592/job/sbt-validator/parambuild?sbt=pull/4617/head&zinc=develop&io=develop&librarymanagement=develop&util=develop&deploy_to_bintray=false&run_sbt_community=false">(restart)</A>

dwijnand commented:
```
[sbt212] [error] /home/jenkinssbt/workspace/sbt-validator/target-0.9.16/project-builds/sbt212-7282eaf24b43dd9e598d0d45a868acfe5552410f/internal/util-complete/src/main/scala/sbt/internal/util/LineReader.scala:90:26: not found: value ScrollUp
[sbt212] [error]     if (showProgress) s"$ScrollUp$DeleteLine" + prompt
[sbt212] [error]                          ^
[sbt212] [error] /home/jenkinssbt/workspace/sbt-validator/target-0.9.16/project-builds/sbt212-7282eaf24b43dd9e598d0d45a868acfe5552410f/internal/util-complete/src/main/scala/sbt/internal/util/LineReader.scala:90:26: not found: value ScrollUp
[sbt212] [error]     if (showProgress) s"$ScrollUp$DeleteLine" + prompt
[sbt212] [error]                          ^
[sbt212] [info] No documentation generated with unsuccessful compiler run
[sbt212] [error] one error found
[sbt212] [error] one error found
[sbt212] [error] (completeProj / Compile / compileIncremental) Compilation failed
[sbt212] [error] (completeProj / Compile / doc) Scaladoc generation failed
```

weird

typesafe-tools commented:
A validation involving this pull request <a href="https://jenkins.scala-sbt.org:8592/job/sbt-validator/1555/console">is in progress...</a>

dwijnand commented:
Rebased and re-ran sbt-validator.

typesafe-tools commented:
[//]: # (Refs: pull/4617/head-develop-develop-develop-develop)
The validator <A HREF="https://jenkins.scala-sbt.org:8592/job/sbt-validator/1555/">has checked</A> the following projects, tested using dbuild, projects built on top of each other.

Project             | Reference            | Commit
:------------------ | :------------------- | :----------------
  sbt               | <A HREF="https://github.com/sbt/sbt/pull/4617">pull/4617/head      </A> | sbt/sbt@15e9356e27958f669c04e8140031652c1e5d8738
  zinc              | <A HREF="https://github.com/sbt/zinc/tree/develop">develop             </A> | sbt/zinc@a5c269c729100140981e1c9ea49568a256fa92f2
  io                | <A HREF="https://github.com/sbt/io/tree/develop">develop             </A> | sbt/io@dda58a019335c1d8480a516bdd1e97ede0b7bdcb
  librarymanagement | <A HREF="https://github.com/sbt/librarymanagement/tree/develop">develop             </A> | sbt/librarymanagement@f073dde812848ef443619578f55e290f3863bcaa
  util              | <A HREF="https://github.com/sbt/util/tree/develop">develop             </A> | sbt/util@3568b13a2e257f33957dd32c51c56ce9c378c9f9

:x: The result is: ***FAILED***
<A HREF="https://jenkins.scala-sbt.org:8592/job/sbt-validator/parambuild?sbt=pull/4617/head&zinc=develop&io=develop&librarymanagement=develop&util=develop&deploy_to_bintray=false&run_sbt_community=false">(restart)</A>

dwijnand commented:
🙄

```
[sbt213-io] [warn] 	::::::::::::::::::::::::::::::::::::::::::::::
[sbt213-io] [warn] 	::          UNRESOLVED DEPENDENCIES         ::
[sbt213-io] [warn] 	::::::::::::::::::::::::::::::::::::::::::::::
[sbt213-io] [warn] 	:: org.scalatest#scalatest_2.13.0-M5;3.0.8-RC2: not found
[sbt213-io] [warn] 	::::::::::::::::::::::::::::::::::::::::::::::
```

The rest of it was successful:

```
[info] Project sbt210-io-------------------: SUCCESS (project rebuilt ok)
[info] Project sbt210-util-----------------: SUCCESS (project rebuilt ok)
[info] Project sbt210-zinc-----------------: SUCCESS (project rebuilt ok)
[info] Project sbt211-io-------------------: SUCCESS (project rebuilt ok)
[info] Project sbt211-librarymanagement----: SUCCESS (project rebuilt ok)
[info] Project sbt211-util-----------------: SUCCESS (project rebuilt ok)
[info] Project sbt211-zinc-----------------: SUCCESS (project rebuilt ok)
[info] Project sbt212----------------------: SUCCESS (project rebuilt ok)
[info] Project sbt212-io-------------------: SUCCESS (project rebuilt ok)
[info] Project sbt212-librarymanagement----: SUCCESS (project rebuilt ok)
[info] Project sbt212-util-----------------: SUCCESS (project rebuilt ok)
[info] Project sbt212-util-interface-------: SUCCESS (project rebuilt ok)
[info] Project sbt212-zinc-----------------: SUCCESS (project rebuilt ok)
[info] Project sbt213-io-------------------: FAILED (RuntimeException: )
[info] Project sbt213-zinc-----------------: SUCCESS (project rebuilt ok)
[info] Project scala-210-------------------: SUCCESS (unchanged, not rebuilt)
[info] Project scala-211-------------------: SUCCESS (unchanged, not rebuilt)
[info] Project scala-212-------------------: SUCCESS (unchanged, not rebuilt)
[info] Project scala-213-------------------: SUCCESS (unchanged, not rebuilt)
[info] >>> The dbuild result is----------------: FAILED (failed: sbt213-io)
```

Ok, I'll recover the commit history now.

eed3si9n commented:
```
[info] Project sbt213-io-------------------: FAILED (RuntimeException: )
```

This is likely due to sbt-validator using Scala 2.13.0-M5 instead of RC1.

eed3si9n commented:
https://github.com/sbt/sbt-validator/commit/7071546c9344a5ff29ea34520a51998e04f89635

typesafe-tools commented:
A validation involving this pull request <a href="https://jenkins.scala-sbt.org:8592/job/sbt-validator/1556/console">is in progress...</a>

typesafe-tools commented:
[//]: # (Refs: pull/4617/head-develop-develop-develop-develop)
The validator <A HREF="https://jenkins.scala-sbt.org:8592/job/sbt-validator/1556/">has checked</A> the following projects, tested using dbuild, projects built on top of each other.

Project             | Reference            | Commit
:------------------ | :------------------- | :----------------
  sbt               | <A HREF="https://github.com/sbt/sbt/pull/4617">pull/4617/head      </A> | sbt/sbt@15e9356e27958f669c04e8140031652c1e5d8738
  zinc              | <A HREF="https://github.com/sbt/zinc/tree/develop">develop             </A> | sbt/zinc@a5c269c729100140981e1c9ea49568a256fa92f2
  io                | <A HREF="https://github.com/sbt/io/tree/develop">develop             </A> | sbt/io@dda58a019335c1d8480a516bdd1e97ede0b7bdcb
  librarymanagement | <A HREF="https://github.com/sbt/librarymanagement/tree/develop">develop             </A> | sbt/librarymanagement@f073dde812848ef443619578f55e290f3863bcaa
  util              | <A HREF="https://github.com/sbt/util/tree/develop">develop             </A> | sbt/util@3568b13a2e257f33957dd32c51c56ce9c378c9f9

:white_check_mark: The result is: ***SUCCESS***
<A HREF="https://jenkins.scala-sbt.org:8592/job/sbt-validator/parambuild?sbt=pull/4617/head&zinc=develop&io=develop&librarymanagement=develop&util=develop&deploy_to_bintray=false&run_sbt_community=false">(restart)</A>

dwijnand committed:
Merge remote-tracking branch 'zinc-lm-intergration/bring-back-a-piece-of-zinc' into zinc-lm-integration

* zinc-lm-intergration/bring-back-a-piece-of-zinc: (90 commits)
  Add scala-integration resolver
  call it compiler-bridge_2.13
  bump Scala bridge to use 2.13.0-RC1
  Break the dependency on LM, and thus Ivy
  Skip Scala 2.10 and Scala 2.11 components for JDK 11
  Refactor compiler bridge unit test
  Fix the several typos detected by github.com/client9/misspell
  Got rid of multiple warnigns
  include only scala-library into the boot classpath during run
  Drop "final" modifier from private case classes, to prevent false positive. Refer https://issues.scala-lang.org/browse/SI-4440
  Support scala 2.13.0-pre-* & 2.13.0-M1
  Split compiler bridge tests to another subproject
  Remove unused imports + variables
  Add back, re-configure & re-enable Scalafmt
  Use Scala 2.12.3
  Remove any reference to `F0` and `F1`
  Adapt Zinc to use the new LibraryManagement API
  Fix ScalaFmt wiring
  Add scalaCompiler variant that accepts ModuleID
  Bump all the modules
  ...

dwijnand commented:
Ready for review, @eed3si9n

eed3si9n committed:
Merge branch 'develop' into pr/4617

eed3si9n committed:
Zinc 1.3.0-M4

eed3si9n committed:
formatting

eed3si9n commented:
Zinc 1.3.0-M4 is out. See https://github.com/dwijnand/sbt/pull/2

dwijnand committed:
Zinc 1.3.0-M4

Zinc 1.3.0-M4

eed3si9n commented on zinc-lm-integration/src/main/scala/sbt/internal/inc/ZincComponentCompiler.scala:
```diff
@@ -0,0 +1,364 @@
+        case sc if (sc startsWith "2.13.0-pre-") => "compiler-bridge_2.13.0-M5"
+        case sc if (sc startsWith "2.13.0-M")    => "compiler-bridge_2.13.0-M5"
+        case sc if (sc startsWith "2.13.0-RC")   => "compiler-bridge_2.13.0-M5"
+        case _                                   => "compiler-bridge_2.13"
```
should be

```
        case "2.13.0-M1"                   => "compiler-bridge_2.12"
        case _                             => "compiler-bridge_2.13"
```

https://github.com/sbt/zinc/pull/651

dwijnand committed:
Switch back to unsuffixed 2.13 compiler bridge sources

Using the newer zinc version there now is a compiler bridge source jar
with suffix _2.13.

dwijnand commented:
Thank you, Eugene, for helping me land this.
